internal/metrics/info.go

```diff
@@ -0,0 +1,150 @@
+
+	// Metrics that may/must be measured or set up before the config is fully parsed
+	// and metric views are initialized
+	buildInfo      *metric.Int64Gauge
+	policyCount    *metric.Int64DerivedGauge
+	configChecksum *metric.Int64Gauge
+	metricRegistry = metric.NewRegistry()
+
+	// ConfigLastReloadView contains the timestamp the configuration was last reloaded, labeled by service
+	ConfigLastReloadView = &view.View{
+		Name:        configLastReload.Name(),
+		Description: configLastReload.Description(),
+		Measure:     configLastReload,
+		TagKeys:     []tag.Key{keyService},
+		Aggregation: view.LastValue(),
+	}
+
+	// ConfigLastReloadSuccessView contains the result of the last configuration reload, labeled by service
+	ConfigLastReloadSuccessView = &view.View{
+		Name:        configLastReloadSuccess.Name(),
+		Description: configLastReloadSuccess.Description(),
+		Measure:     configLastReloadSuccess,
+		TagKeys:     []tag.Key{keyService},
+		Aggregation: view.LastValue(),
+	}
+)
+
+// SetBuildInfo records the pomerium build info.  You must call RegisterInfoMetrics to
+// have this exported
+func SetBuildInfo(service string) {
+	if buildInfo == nil {
+		buildInfoL, err := metricRegistry.AddInt64Gauge("build_info",
+			metric.WithDescription("Build Metadata"),
+			metric.WithLabelKeys("service", "version", "revision", "goversion"),
+		)
+		if err != nil {
+			log.Error().Err(err).Msg("internal/metrics: failed to register build info metric")
+			return
+		}
+		buildInfo = buildInfoL
+	}
+
+	m, err := buildInfo.GetEntry(
+		metricdata.NewLabelValue(service),
+		metricdata.NewLabelValue(version.FullVersion()),
+		metricdata.NewLabelValue(version.GitCommit),
+		metricdata.NewLabelValue((runtime.Version())),
+	)
+	if err != nil {
+		log.Error().Err(err).Msg("internal/metrics: failed to add build info metric")
+	}
+	m.Set(1)
+}
+
+// SetConfigInfo records the status, checksum and timestamp of a configuration reload.  You must register InfoViews or the related
+// config views before calling
+func SetConfigInfo(service string, success bool, checksum string) {
+
+	if success {
+		serviceTag := tag.Insert(keyService, service)
+		if err := stats.RecordWithTags(
+			context.Background(),
+			[]tag.Mutator{serviceTag},
+			configLastReload.M(time.Now().Unix()),
+		); err != nil {
+			log.Error().Err(err).Msg("internal/metrics: failed to record config checksum timestamp")
+		}
+
+		if err := stats.RecordWithTags(
+			context.Background(),
+			[]tag.Mutator{serviceTag},
+			configLastReloadSuccess.M(1),
+		); err != nil {
+			log.Error().Err(err).Msg("internal/metrics: failed to record config reload")
+		}
+	} else {
+		stats.Record(context.Background(), configLastReloadSuccess.M(0))
+	}
+}
+
+// Register non-view based metrics registry globally for export
+func RegisterInfoMetrics() {
+	metricproducer.GlobalManager().AddProducer(metricRegistry)
```

Comment:
This is still a bit black magic to me; what are the implications of this?

Reply:
This is akin to registering a view.  Views and Metrics are handled a bit differently - the latter has the behavior we want/need for config derived info early on.  

The setup for Metric based stats is in reverse order.  Rather than Measure -> View -> Register you need a Registry and then Registry -> Metric -> Entry.  

We could do this in init() safely, which might address the metric singleton pattern better.  I will elaborate more in those comments.

I added a test to demonstrate that multiple AddProducer() calls with the same registry does not result in multiple references to the same registry internally.

Reply:
> We could do this in init() safely, which might address the metric singleton pattern better. I will elaborate more in those comments.

 I think that makes sense. 

> I added a test to demonstrate that multiple AddProducer() calls with the same registry does not result in multiple references to the same registry internally.

 👍

Reply:
Going to leave this as is so we can opt-in/out of various metrics later.  We're not doing anything in init() to move this next to.